mappers/modbus/device/device.go

```diff
@@ -0,0 +1,251 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package dev
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"regexp"
+	"strconv"
+	"sync"
+	"time"
+
+	mqtt "github.com/eclipse/paho.mqtt.golang"
+	mappercommon "github.com/kubeedge/kubeedge/mappers/common"
+
+	"github.com/kubeedge/kubeedge/mappers/modbus/configmap"
+	"github.com/kubeedge/kubeedge/mappers/modbus/driver"
+	"github.com/kubeedge/kubeedge/mappers/modbus/globals"
+	"k8s.io/klog"
+)
+
+var devices map[string]*globals.ModbusDev
+var models map[string]mappercommon.DeviceModel
+var protocols map[string]mappercommon.Protocol
+var wg sync.WaitGroup
+
+// getDeviceID extract the device ID from Mqtt topic.
+func getDeviceID(topic string) (id string) {
+	re := regexp.MustCompile(`hw/events/device/(.+)/twin/update/delta`)
+	return re.FindStringSubmatch(topic)[1]
+}
+
+// onMessage callback function of Mqtt subscribe message.
+func onMessage(client mqtt.Client, message mqtt.Message) {
+	klog.Info("Receive message", message.Topic())
+	// Get device ID and get device instance
+	id := getDeviceID(message.Topic())
+	if id == "" {
+		klog.Error("Wrong topic")
+		return
+	}
+	klog.Info("Device id: ", id)
+
+	var device *globals.ModbusDev
+	var ok bool
+	if device, ok = devices[id]; !ok {
+		klog.Error("Device not exist")
+		return
+	}
+
+	// Get twin map key as the propertyName
+	var delta mappercommon.DeviceTwinDelta
+	if err := json.Unmarshal(message.Payload(), &delta); err != nil {
+		klog.Error("Unmarshal message failed: ", err)
+		return
+	}
+	for twinName, twinValue := range delta.Delta {
+		i := 0
+		for i = 0; i < len(device.Instance.Twins); i++ {
+			if twinName == device.Instance.Twins[i].PropertyName {
+				break
+			}
+		}
+		if i == len(device.Instance.Twins) {
+			klog.Error("Twin not found: ", twinName)
+			continue
+		}
+		// Type transfer
+		device.Instance.Twins[i].Desired.Value = twinValue
+		var visitorConfig configmap.ModbusVisitorConfig
+		if err := json.Unmarshal([]byte(device.Instance.Twins[i].PVisitor.VisitorConfig), &visitorConfig); err != nil {
+			klog.Error("Unmarshal visitor config failed")
+		}
+		valueConverted, err := mappercommon.Convert(device.Instance.Twins[i].PVisitor.PProperty.DataType, twinValue)
+		if err == nil {
+			valueInt, _ := valueConverted.(int64)
+			device.ModbusClient.Set(visitorConfig.Register, visitorConfig.Offset, uint16(valueInt))
+		} else {
+			klog.Error("Convert failed")
+		}
+	}
+}
+
+// isRS485Enabled is RS485 feature enabled for RTU.
+func isRS485Enabled(customizedValue configmap.CustomizedValue) bool {
+	isEnabled := false
+
+	if len(customizedValue) != 0 {
+		if value, ok := customizedValue["serialType"]; ok {
+			if value == "RS485" {
+				isEnabled = true
+			}
+		}
+	}
+	return isEnabled
+}
+
+// initModbus initialize modbus client
+func initModbus(protocolConfig configmap.ModbusProtocolCommonConfig, slaveID int16, client *driver.ModbusClient) error {
```

Comment:
if protocolConfig.Com.SerialPort and protocolConfig.Tcp.Ip are both null or both not null,it's useless,I think it's worthy checking
```
// tcp and rtu are mutually exclusive
	if pcc.Com.SerialPort != "" && pcc.Tcp.Ip != "" {
		klog.Error("tcp and rtu are mutually exclusive.")
		return fmt.Errorf("tcp and rtu are mutually exclusive.")
	}

	if pcc.Com.SerialPort == "" && pcc.Tcp.Ip == "" {
		klog.Error("SerialPort and Ip are both null")
		return fmt.Errorf("tcp and rtu are mutually exclusive.")
	}

```